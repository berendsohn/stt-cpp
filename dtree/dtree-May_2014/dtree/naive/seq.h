// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_SEQ_H_
#define DTREE_NAIVE_SEQ_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndSeq : public Base {
 public:
  EndSeq() : solid_() {}
  EndSeq* solid(int i) const { return solid_[i]; }
  void set_solid(int i, EndSeq* solid_i) { solid_[i] = solid_i; }
  template<typename Forwarder>
  void CopyFrom(const EndSeq& node, Forwarder forward);
 private:
  EndSeq* solid_[2];
  DISALLOW_COPY_AND_ASSIGN(EndSeq);
};

template<typename Group, typename Base>
class WithValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToSeq(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromSeq(Node* node, typename Group::Type delta);

 protected:
  WithValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  template<typename Node>
  static typename Semigroup::Type AggrSeq(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostSeq(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeftmostSeq(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRightmostSeq(Node* node, const Predicate& predicate);
 protected:
  WithAggr() {}
};

template<typename Base>
class WithReverseBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverseBy() {}
};

template<typename Base>
class WithReverse : public Base {
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverse() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_SEQ_H_
