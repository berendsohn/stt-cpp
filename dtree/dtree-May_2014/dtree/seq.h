// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_SEQ_H_
#define DTREE_SEQ_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/seq.h"
#else
#include "dtree/selfadjust/seq.h"
#endif
#endif  // DTREE_SEQ_H_
