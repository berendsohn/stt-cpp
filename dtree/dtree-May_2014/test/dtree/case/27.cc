#define DTREE_CONFIG EndSeq<WithAggr<FreeSemigroup, NullPredicate_, WithValue<FreeGroup, WithValue<FreeGroup, Begin > > > >
#include "test/dtree/seq.cc"
