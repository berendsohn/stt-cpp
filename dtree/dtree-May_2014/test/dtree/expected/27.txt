<<<
  15: v790,000,e9f,000 vfea,000,5d5,000
  3: vf2f,000,cd1,000 v89f,000,6ce,000
  9: v88d,000,e5c,000 vd33,000,da9,000
  2: v956,000,524,000 veef,000,592,000
  11: vc9a,000,b83,000 v40e,000,326,000
  1: vb65,000,a73,000 vcb3,000,24a,000
  4: v75f,000,011,000 v873,000,683,000
  6: vb4f,000,499,000 v049,000,549,000
  13: vf47,000,6f9,000 v00a,000,5ea,000
  12: ve0c,000,0f0,000 ve69,000,875,000
  8: v0c4,000,205,000 v8ff,000,c0e,000
  10: v9af,000,dbc,000 vb13,000,725,000
  5: v1bf,000,e9c,000 vac5,000,550,000
  7: v37c,000,83a,000 veb7,000,66a,000
  14: vc45,000,45d,000 v790,000,b3f,000
>>>
<<<
  @15: v790,000,e9f,000 vfea,000,5d5,000 a000,000
      /3: vf2f,b4f,cd1,499 v89f,049,6ce,549 a89f,6ce
          /9: v88d,75f,e5c,011 vd33,873,da9,683 ad33,da9
            \2: v956,88d,524,e5c veef,d33,592,da9 aeef,592
                /11: vc9a,b65,b83,a73 v40e,cb3,326,24a a40e,326
              \1: vb65,956,a73,524 vcb3,eef,24a,592 acb3,24a
        \4: v75f,f2f,011,cd1 v873,89f,683,6ce a873,683
    \6: vb4f,790,499,e9f v049,fea,549,5d5 a049,549
        /13: vf47,37c,6f9,83a v00a,eb7,5ea,66a a00a,5ea
            /12: ve0c,1bf,0f0,e9c ve69,ac5,875,550 ae69,875
              \8: v0c4,e0c,205,0f0 v8ff,e69,c0e,875 a8ff,c0e
                \10: v9af,0c4,dbc,205 vb13,8ff,725,c0e ab13,725
          \5: v1bf,f47,e9c,6f9 vac5,00a,550,5ea aac5,550
      \7: v37c,b4f,83a,499 veb7,049,66a,549 aeb7,66a
        \14: vc45,37c,45d,83a v790,eb7,b3f,66a a790,b3f
>>>
